README.md
# ForteXorcist

[![Build Status](https://travis-ci.org/bellexandria/ForteXorcist.svg?branch=master)](https://travis-ci.org/bellexandria/ForteXorcist)

I did not write this initial addon, but its one of my favourites. This project is based on ForteXorcist 1.980.8 by Xussie-Kazzak [EU].

## Needs work
The following classes/specs need to be checked for errors and spells:

- [ ] Core config pages
- [ ] Warlock
  - [ ] Affliction
  - [ ] Destruction
  - [ ] Demonology

- [ ] Warrior
  - [ ] Protection
  - [ ] Fury
  - [ ] Arms
- [ ] Druid
  - [ ] Feral
  - [ ] Guardian
  - [ ] Balance
  - [ ] Restoration
- [ ] Hunter
  - [ ] Beast Mastery
  - [ ] Marksmanship
  - [ ] Survival
- [ ] Paladin
  - [ ] Protection
  - [ ] Holy
  - [ ] Retribution
- [ ] Rogue
  - [ ] Assassination
  - [ ] Outlaw
  - [ ] Subtlety
- [ ] Priest
  - [ ] Holy
  - [ ] Shadow
  - [ ] Discipline
- [ ] Mage
  - [ ] Arcane
  - [ ] Fire
  - [ ] Frost
- [ ] Death Knight
  - [ ] Blood
  - [ ] Frost
  - [ ] Unholy
- [ ] Demon Hunter
  - [ ] Vengeance
  - [ ] Havoc
- [ ] Shaman
  - [ ] Elemental
  - [ ] Enhancement
  - [ ] Restoration
- [ ] Monk
  - [ ] Brewmaster
  - [ ] Mistweaver
  - [ ] Windwalker


CHANGELOG.md
# Changelog

## Unreleased

- Added the Demon Hunter Module
- Updates for 7.3.5 Legion patch compatability
- Fixed sound channels for new api
- Updated Classes for 8.0.0 Battle for Azeroth pre-patch

## Legacy Releases

### [v1.980.8] - 2012-09-27

- Rogue: Updated class module.
- Warrior: Updated class module.

### [v1.980.7]

- Monk: Added class module.
- Rogue: Updated class module. I'm still working on adding MoP spells.
- Spell Timer: Fixed bug related to spells that now have a self buff ánd a target debuff. I will investigate this problem further for the next update.

### [v1.980.6]

- Paladin: Updated class module.

### [v1.980.5]

- Druid: Updated class module.

### [v1.980.4]

- Hunter: Updated class module.
- Shaman: Updated class module.
- Warlock: Updated class module.
- General: Some fixes to the Options panel. You can now scroll all the way down again. Option tooltips are now loaded differently, so that you should never get a 'script ran too long' error.

### [v1.980.3]

- Mage: Updated class module.

### [v1.980.2]

- Priest: Updated class module.
- Warlock: Updated class module.

### [v1.980.1]

- Death Knight: Updated class module. Spells should now be more correct and complete.
- Warlock: Updated class module. Spells should now be more correct and complete.

### [v1.98]

- General: Works with Patch 5.0 / Mists of Pandaria.
- General: There are still lots of spells that need to be (re-)added, but at least it shouldn't give any lua errors. I will try to go through all classes in the coming weeks.
- General: The option to enable the display of registered spells to spell type groups is moved to 'Advanced Options > Options Behavior > List spells on mouse-over', because this now also applies to the Cooldown Timer.

### [v1.976.3]

- General: Added some internal/hidden cooldowns.
- Spell Timer: Fixed bug in de update of some Cooldowns on the Spell Timer, like Death Knight's Runes changing into Death Runes.
- Cooldown Timer: Internal/hidden cooldowns are now added to their own cooldown group, separate from Powerups.
- Cooldown Timer: You can now see what spells are hard-coded to the cooldown groups too (Advanced Options > Spell Timer > List spells on mouse-over).

### [v1.976.2]

- General: Hiding the tooltips on the Spell Timer works again.
- General: Removed old/dummy class modules from the addon package. You can safely delete all the class modules (Forte_DeathKnight, Forte_Druid, Forte_Hunter, Forte_Mage, Forte_Paladin, Forte_Priest, Forte_Rogue, Forte_Shaman, Forte_Vehicle, Forte_Warlock and Forte_Warrior). Make sure you keep and enable the Forte_Class module.
- General: Many spell ids across all modules have been corrected or removed.
- Spell Timer: You can see what spells belong to the spell type groups by mousing over their options. For real this time! ;) This is not enabled by default, so you have to go to 'Advanced Options > Spell Timer > List spells on mouse-over' to switch it on.
- Cooldown Timer: changed hidden cooldown of Lightweave to 65 seconds (from 45).

### [v1.976]/[v1.976.1]

- Spell Timer: Spells are now added to the class modules in a different way, which allows for much more flexibility. While this is super nice, I can't guarantee that everything still works as it should, simply because I can't test a lot myself.
- Spell Timer: Fixed many small bugs and features. E.g. all spells can now show their damage/healing ticks on the bars, not just damage-over-time spells.
- Spell Timer: You can now see Weapon enchants on the Spell Timer too.
- Spell Timer: You can see what spells belong to the spell type groups by mousing over their options.
- Spell Timer: Changed the spell types from their 'technical properties' to their practical use (e.g. crowd control type). I will further build on this in the next versions.
- Spell Timer: Mousing over Spell Timer icons will now also show the 'real' tooltip for the (de)buff or spell.
- Cooldown Timer: Ranged weapon enchants are now shown too.
- General: Mousing over a default button will now display the default in the option itself, instead of in the tooltip.
- General: Added a fix that allows '%' signs in Party/Raid messages, without giving errors. Before this fix only properly formatted text, like %% or %s was possible. This should at least fix the Recklessness message bug.
- General: The colour black (0.00 0.00 0.00) will now also work properly on 64-bit clients.
- IMPORTANT: All Class modules are now part of the Forte_Class Module. This will avoid problems because of class modules not being enabled. This will not impact the in-game performance. THE OLD CLASS MODULE FILES WILL REMAIN IN THE PACKAGE AS EMPTY DUMMIES FOR THE NEXT COUPLE OF UPDATES, TO AVOID MANUAL UPDATE PROBLEMS.

### [v1.975.9]

- Spell Timer: Fixed bug in channel timer updates. The duration would be wrong if you refreshed a channelled spell. When you get 'interrupted', the duration of channels is still more accurate than in the default WoW interface.
- Spell Timer: Fixed bug with timers without a real duration not being removed properly. 'Raid debuffs' were not affected by this bug.
- Warrior: Added a number of announcement messages to the Warrior class module. Thanks to KrazyKrl!
- General: Added a number of spells to various modules. This should be most of the requested spells from the last five months (that are possible at the moment).
- General: 'yell' is now also a valid output channel for the Party/Raid messages.
- General: I removed some compatibility fixes. Now you can no longer use the global frame names FX_Timer, FX_Cooldown and FX_Splash. Instead, just use the names FX_Timer1, FX_Cooldown1 and FX_Splash1.
- General: If you have problems with the font of the options interface in your language, you can use the command "/fx resetfont" to set it to a safe default.

### [v1.975.8]

- Version update to 4.3.0.
- Added some spells to the timer modules, but I will add more in the next update.

### [v1.975.7]


Forte_Class/Shaman/Shaman.lua
--ForteXorcist v1.980.8 by Xus 28-09-2012 for 5.0

if FW.CLASS == "SHAMAN" then
  local FW = FW;
  local FWL = FW.L;
  local SH = FW:ClassModule("Shaman");

  local ST = FW:Module("Timer");
  local CA = FW:Module("Casting");
  local CD = FW:Module("Cooldown");

  -- Register ID renames first!


  if ST then
    local F = ST.F;
    ST:SetDefaultHasted(1)

    :AddChannel(61882, 0, 1) -- Earthquake

    -- Elemental Spells
    :AddSpell( 8050,  21,"Default",  F.TICKS) -- Flame Shock
    :AddSpell( 8056,   8,"Default") -- Frost Shock

    -- Enhancement Spells
    :AddSpell(51533,  30,"Pet",    F.SUMMON) -- Feral Spirit
    :AddSpell(17364,  12,"Default") -- Stormstrike
    -- Restoration Spells
    --:AddSpell(85838,  15,"Buff",  F.BUFF) -- Ancestral Fortitude
    :AddSpell(  974, 600,"Buff",  F.BUFF+F.UNIQUE) -- Earth Shield
    :AddSpell(76780,  50,"Crowd",  F.UNIQUE) -- Bind Elemental
    :AddSpell(51945,  12,"Heal",  F.HOT) -- Earthliving
    :AddSpell(61295,  12,"Heal",  F.HOT) -- Riptide

    :AddSpell(61882,  10,"Default",  F.AOE_DMG):SetTickSpeed(1) -- Earthquake

    :AddSpell(73920,  10,"Heal",  F.AOE_HOT):SetTickSpeed(2) -- Healing Rain

    :AddCooldown(51505,  8) -- Lava Burst
    :AddCooldown(51490, 45) -- Thunderstorm
    :AddCooldown(  421,  6) -- Chain Lightning

    :SetTickSpeed(2,8190) -- Magma Totem
    --:SetTickSpeed(2,25858) -- Stoneclaw Totem
    --:SetTickSpeed(3, 3814) -- Tremor Totem

    -- Buffs
    :AddBuff(108271) -- Astral Shift
    :AddBuff(108281) -- Ancestral Guidance
    :AddBuff(16188) -- Ancestral Swiftness
    :AddBuff(79206) -- Spiritwalker's Grace
    :AddBuff(58875) -- Spirit Walk
    :AddBuff(16278) -- Flurry

    -- Elemental
    :AddBuff(324) -- Lightning Shield
    :AddBuff(16166) -- Elemental Mastery
    :AddBuff(16246) -- Clearcasting

    -- Enhancement
    :AddBuff(53817) -- Maelstrom Weapon
    :AddBuff(30823) -- Shamanistic Rage


    -- Restoration
    :AddBuff(53390) -- Tidal Waves
    :AddBuff(52128) -- Water Shield

    :AddCasterBuffs()
    :AddMeleeBuffs()

    do
      -- Code to track totems
      local select = select;
      local strfind = strfind;
      local index_to_group = {"TotemFire","TotemEarth","TotemWater","TotemAir"};
      local function SH_TotemUpdate(event,index)
        if not index or index < 1 or index > 4 then return; end
        -- Fire = 1 Earth = 2 Water = 3 Air = 4
        local _, name, startTime, duration, icon = GetTotemInfo(index);
        local group = index_to_group[index];
        local i = ST.ST:find(group,21);
        if i then
          if name ~= "" then
            ST.ST:remove(i);
          else
            ST:Fade(i,(ST.ST[i][1]-GetTime()<0.75) and 2 or 3);
          end
        end
        if name ~= "" then
          ST:AddManualSpellOfType(name,ST.DEFAULT);
          ST.ST:insert(startTime+duration,0,duration,name,0,ST.DEFAULT,icon,name,0,0,"none",0,ST.PRIOR_NONE,0,1,0,0,00000,0,startTime+duration,group,1.0,0,0,0,1,"",0);
        end
      end
      FW:RegisterToEvent("PLAYER_TOTEM_UPDATE", SH_TotemUpdate);

      FW:RegisterDelayedLoadEvent(function(self)
        for i=1,4,1 do
          SH_TotemUpdate(self,i);
        end
      end);
    end

    FW:SetMainCategory(FWL.SPELL_TIMER):SetSubCategory(FWL.MY_SPELLS)
      :AddOption("CO2",FWL.TOTEM_FIRE,  "",  "TotemFire",99):SetFunc(ST.FilterChange)
      :AddOption("CO2",FWL.TOTEM_EARTH,  "",  "TotemEarth",99):SetFunc(ST.FilterChange)
      :AddOption("CO2",FWL.TOTEM_WATER,  "",  "TotemWater",99):SetFunc(ST.FilterChange)
      :AddOption("CO2",FWL.TOTEM_AIR,    "",  "TotemAir",99):SetFunc(ST.FilterChange)

    FW.InstanceDefault.Timer.TotemFire =   {[0]=true,1.00,0.37,0.00};
    FW.InstanceDefault.Timer.TotemEarth =   {[0]=true,1.00,0.56,0.00};
    FW.InstanceDefault.Timer.TotemWater =   {[0]=true,0.00,1.00,0.67};
    FW.InstanceDefault.Timer.TotemAir =   {[0]=true,0.00,1.00,1.00};
  end

  if CD then
    CD:AddCasterPowerupCooldowns();
    CD:AddMeleePowerupCooldowns();
  end
end
